core/services/pipeline/runner.go

```diff
@@ -176,22 +183,31 @@ func NewRun(spec Spec, vars Vars) Run {
 	}
 }
 
+func (r *runner) OnRunFinished(fn func(*Run)) {
+	r.runFinished = fn
+}
+
 func (r *runner) ExecuteRun(
 	ctx context.Context,
 	spec Spec,
 	vars Vars,
 	l logger.Logger,
 ) (Run, TaskRunResults, error) {
-	l.Debugw("Initiating tasks for pipeline run of spec", "job ID", spec.JobID, "job name", spec.JobName)
-
 	run := NewRun(spec, vars)
 
-	taskRunResults, err := r.run(ctx, &run, vars, l)
+	pipeline, err := r.initializePipeline(&run)
+
+	if err != nil {
+		return run, nil, err
+	}
+
+	taskRunResults, err := r.run(ctx, pipeline, &run, vars, l)
 	if err != nil {
 		return run, nil, err
 	}
 
-	if run.Async && run.Pending {
+	// TODO: do this HasAsync check earlier
```

Comment:
Why?

Reply:
ExecuteRun is the legacy codepath that doesn't support async. If I check for async tasks before even starting the pipeline, that could save us from running a pipeline that will then fail here anyway. ExecuteRun is becoming less and less relevant though and almost no code paths use it anymore, so I'm considering leaving it as is